src/include/mempool.h

```diff
@@ -218,7 +218,7 @@ static size_t pick_a_shard_int() {
   // Dirt cheap, see:
   //   https://fossies.org/dox/glibc-2.32/pthread__self_8c_source.html
   size_t me = (size_t)pthread_self();
-  size_t i = (me >> CEPH_PAGE_SHIFT) & ((1 << num_shard_bits) - 1);
+  size_t i = (me >> 16) & ((1 << num_shard_bits) - 1);
```

Comment:
`pthread_self()`-as-shard-selector is a really bad idea. I would prefer to switch to the way we already use in crimson – please take a look on these PRs and the discussion there:

* https://github.com/ceph/ceph/pull/53130
* https://github.com/ceph/ceph/pull/53245

Reply:
It is different from crimson. Crimson is shard per core, just one thread running on the core until the end, so crimson can use the `sched_getcpu()`. For classic osd, there are so many threads (tp_osd_tp, msgr_worker,kvthread), running on the cpus in a shared way, use the `sched_getcpu()` may be cause more cache ping-pongs between cpu cores.

Reply:
If _GNU_SOURCE is defined:

The crimson patch calls sched_getcpu() everytime mempool::adjust_count is called which means a given CPU core will choose the same shard (regardless of which thread happens to be running on that core at the time). In the case of crimson the thread affinity is bound to the core so there is no chance it will switch cores between the call to sched_getcpu() and the increment.

For classic there is a (tiny) risk of a context switch and core hop between calling sched_getcpu() and the increment. The use of atomics means this exceptional case is still correct, just with non-optimal performance. So the patch would work fine for classic.

If _GNU_SOURCE is not defined:

The crimson patch calls sched_getcpu() once to assign an ID to the thread and then uses this every time mempool::adjust_count is called. As the thread affinity is bound to a core this works well for crimson and avoids shards from being accessed by different cores.

For classic threads aren't bound to cores so this won't work very well, although it might do a better job that the current arithmetic on pthread_self. A better fix for classic in this scenario would be to use a global variable that is incremented modulo number of shards to select a shard the first time a thread calls mempool::adjust_count and then stick to this shard on subsequent calls. If threads switch cores this won't prevent CPU cache lines bouncing between cores (but then the current pthread_self() implementation doesn't do this either) but it will at least guarantee a perfect distribution of threads across shards which is better than the current implementation or the proposed change

Reply:
> For classic there is a (tiny) risk of a context switch and core hop between calling sched_getcpu() and the increment. The use of atomics means this exceptional case is still correct, just with non-optimal performance. So the patch would work fine for classic.

Yup, that's the exact justification for the atomics there. I think we could squeeze them **if** we could guarantee there is no way to hop between cores (everything pinned?). I don't know whether this is a real problem.